accord-core/src/test/java/accord/topology/TopologyRandomizer.java

```diff
@@ -365,4 +367,17 @@ private boolean everyShardHasOverlaps(long prevEpoch, Shard[] in, Shard[] out)
         Invariants.checkState (i == in.length && o == out.length);
         return true;
     }
+
+    public void onStale(Node.Id id, Timestamp sinceAtLeast, Ranges ranges)
```

Comment:
So for test purposes we don't expect there to be any stale ranges in the protocol? Which makes sense of course since we haven't added anything to allow nodes to fall behind and re-bootstrap yet.

Reply:
We do expect stale ranges, but only when we lose a range and haven't finished catching up. We should add chaos to the burn test for permitting a node go stale otherwise, but we need to control this to ensure we do not lose quorum.

Reply:
OK, one point of confusion is that if stale ranges occur this generates an exception which would cause a test failure?

I modifed `onStale` to always throw `IllegalStateException` and everything seems to pass.

When a node loses a range why would it be stale at the node losing the range while attempting to recover a transaction? Shouldn't it be known at that point that the transaction does not apply to this node?

Or is it the gaining node where it is stale because it was unable to get the transactions after it bootstrapped the new range to apply. How does that happen with transaction metadata in each epoch and other nodes currently not truncating transactions until they are universally durable?

Reply:
This will only fail the burn test, and is not guaranteed to fail every run as it is a rare occurrence.

If you have a replica set of {1,2,3} in epoch 1 that changes to {1,2,4} in epoch 2, and then before {3} has caught up with everything that happened in epoch 1, {1,2,4} all CoordinateShardDurable and mark some portion of the range owned by {3} in epoch 1 as durable, they begin cleaning up their state associated with these transactions - since {3} wasn't a party to the CoordinateShardDurable it wasn't asked if it was up to date.

We could probably include {3} also in any CoordinateShardDurable for ranges not known to have retired earlier epochs for. But we have to handle staleness anyway.